pageserver/src/consumption_metrics/disk_cache.rs

```diff
@@ -0,0 +1,66 @@
+use anyhow::Context;
+use std::path::PathBuf;
+use std::sync::Arc;
+use tempfile::NamedTempFile;
+
+use super::RawMetric;
+
+pub(super) async fn read_metrics_from_disk(path: Arc<PathBuf>) -> anyhow::Result<Vec<RawMetric>> {
+    // do not add context to each error, callsite will log with full path
+    let span = tracing::Span::current();
+    tokio::task::spawn_blocking(move || {
+        let _e = span.entered();
+        let mut file = std::fs::File::open(&*path)?;
+        let reader = std::io::BufReader::new(&mut file);
+        anyhow::Ok(serde_json::from_reader::<_, Vec<RawMetric>>(reader)?)
+    })
+    .await
+    .context("read metrics join error")
+    .and_then(|x| x)
+}
+
+pub(super) async fn flush_metrics_to_disk(
+    current_metrics: &Arc<Vec<RawMetric>>,
+    final_path: &Arc<PathBuf>,
+) -> anyhow::Result<()> {
+    use std::io::Write;
+
+    anyhow::ensure!(
+        final_path.parent().is_some(),
+        "path must have parent: {final_path:?}"
+    );
+
+    let span = tracing::Span::current();
+    tokio::task::spawn_blocking({
```

Comment:
Perhaps add a comment to explain why we're doing sync IO in a spawn_blocking() instead of using `tokio::fs`.  I'm guessing the main motivation is to enable using the sync NamedTempFile?

Reply:
The motivation is that all of the work is blocking work.

When you use `tokio::fs`, then you are doing this `spawn_blocking` dance just multiple times. I think this is quite natural use of `spawn_blocking`, so unsure why there should be a comment.

Reply:
Not objecting to using spawn_blocking here.  However in an async codebase it's always a bit noteworthy when we do some sync I/O, so can helpful to leave a comment explaining why.  It's a habit of mine, from working in environments where one little stray sync blocking call could ruin everyone's day :-)

(True that tokio::fs doing nothing more clever than spawn_blocking, but as we adopt io_uring the penalty for doing a few independent async I/O calls vs. doing one wrapped spawn_blocking will be lower.)